R/get_ci.R

```diff
@@ -83,95 +167,297 @@ get_ci <- function(dist, interval, estimate, variance = NULL, denominator = NULL
     stop("`estimate` must be numeric")
   }
 
-  if (dist == "poisson") {
-    if (!all(estimate %% 1 == 0, na.rm = TRUE)) {
-      stop("`estimate must be an integer if `dist` is 'poisson'")
-    }
+  if (!is.numeric(variance_log)) {
+    stop("`variance_log` must be numeric")
+  }
 
-    if (!is.null(variance)) {
-      stop("`variance` must be left empty when `dist` is 'poisson'")
-    }
+  # TODO check if variance_log > 0
 
-    if (is.null(denominator)) {
-      denominator <- 1
-    }
-  } else { # normal, log normal
+  if (length(estimate) != length(variance_log)) {
+    stop("length of `variance_log` must be equal to length of `estimate`")
+  }
 
-    if (is.null(variance)) {
-      stop("`variance` is required to construct a confidence interval unless `dist` is 'poisson'")
-    }
+  # define alpha
+  alpha <- 1 - interval
 
-    if (!is.numeric(variance)) {
-      stop("`variance` must be numeric")
-    }
+  # follow SAS implementation
+  result <- data.frame(
+    lower = estimate * exp(-(stats::qnorm(p = 1 - (alpha / 2), mean = 0, sd = 1) * sqrt(variance_log))),
+    upper = estimate * exp(stats::qnorm(p = 1 - (alpha / 2), mean = 0, sd = 1) * sqrt(variance_log))
+  )
 
-    if (!all(variance >= 0, na.rm = TRUE)) {
-      stop("`variance` must be greater than or equal to 0")
-    }
+  return(result)
+}
 
-    if (length(estimate) != length(variance)) {
-      stop("length of `variance` must be equal to length of `estimate`")
-    }
 
-    if (!is.null(denominator)) {
-      stop("`denominator` must be NULL when `dist` is not 'poisson'")
-    }
+#' Calculate a Poisson distribution confidence interval
+#'
+#' @description Calculate a Poisson distribution confidence interval. This
+#' function imitates [stats::qpois()] using [stats::qchisq()] to allow for the
+#' continuous extension of the estimate.
+#'
+#' @param interval A scalar between 0 and 1, indicating the width of the
+#' interval. For example, use `0.95` to calculate a 95% confidence interval.
+#' @param x A numeric vector of positive integers. It can contain the
+#' stratum-specific number of events or the observed number of events in the
+#' study population.
+#' @param y A vector of positive values. It can contain the population-time for
+#' each stratum or the expected number of events. The default value is `1`.
+#'
+#' @details When working with rates, pass the stratum-specific
+#' number of events into `x` and the population-time for each stratum into `y`.
+#' When working with standardized morbidity/mortality ratios, pass the observed
+#' number of events into `x` and the expected number of events into `y`. Ensure
+#' that the values of `x` and `y` are aligned and that the vectors are the same
+#' length.
+#'
+#' To calculate a confidence interval around count data, pass a vector
+#' of counts into `x` and set `y = 1`. The shorter vector `y` will
```

Comment:
I think the wording here should tell the user to not pass any value for `y`, so that the default value is used. Then we are not saying that `x` and `y` must be the same length while also saying it's okay for `y` to be a scalar but only if that scalar is `1`. 

Alternatively, perhaps a more robust way to code this is to set the default value for `y` to be `NULL` and then the function treats this as counts data only if `y` is NULL.